packages/connect-node/src/grpc-http-transport.ts

```diff
@@ -0,0 +1,408 @@
+  // Unary Request
+  req: UnaryRequest<I>;
+
+  // Payload encoding
+  encoding: "utf8" | "binary";
+
+  // Request body
+  payload: Uint8Array;
+}
+
+const messageFlag = 0b00000000;
+
+/**
+ * Create a Transport for the gRPC protocol using the
+ * Node.js `http` or `https package.
+ *
+ */
+export function createGrpcHttpTransport(
+  options: GrpcHttpTransportOptions
+): Transport {
+  const useBinaryFormat = options.useBinaryFormat ?? false;
+  return {
+    async unary<
+      I extends Message<I> = AnyMessage,
+      O extends Message<O> = AnyMessage
+    >(
+      service: ServiceType,
+      method: MethodInfo<I, O>,
+      signal: AbortSignal | undefined,
+      timeoutMs: number | undefined,
+      header: HeadersInit | undefined,
+      message: PartialMessage<I>
+    ): Promise<UnaryResponse<O>> {
+      const { normalize, serialize, parse } = createClientMethodSerializers(
+        method,
+        useBinaryFormat,
+        options.jsonOptions,
+        options.binaryOptions
+      );
+      try {
+        return await runUnary<I, O>(
+          {
+            stream: false,
+            service,
+            method,
+            url: createMethodUrl(options.baseUrl, service, method).toString(),
+            init: {},
+            header: grpcCreateRequestHeader(useBinaryFormat, timeoutMs, header),
+            message: normalize(message),
+            signal: signal ?? new AbortController().signal,
+          },
+          async (req: UnaryRequest<I>): Promise<UnaryResponse<O>> => {
+            const envelope = encodeEnvelope(
+              messageFlag,
+              serialize(req.message)
+            );
+            const encoding = useBinaryFormat ? "binary" : "utf8";
+            const response = await makeNodeRequest({
+              req,
+              payload: envelope,
+              encoding,
+              httpOptions: options.httpOptions,
+            });
+
+            const responseHeaders = nodeHeaderToWebHeader(response.headers);
+            assert(
+              typeof response.statusCode == "number",
+              "http1 client response is missing status code"
+            );
+            validateResponse(
+              useBinaryFormat,
+              response.statusCode,
+              responseHeaders
+            );
+
+            const messageResult = await readEnvelope(response);
+
+            let trailer = mapResponseTrailers(response.trailers);
+            if (receivedTrailers(trailer)) {
+              validateGrpcStatus(trailer);
+            }
+
+            if (messageResult.done) {
+              throw "premature eof";
+            }
+
+            const eofResult = await readEnvelope(response);
+            trailer = mapResponseTrailers(response.trailers);
+            if (receivedTrailers(trailer)) {
+              validateGrpcStatus(trailer);
+            }
+
+            if (!eofResult.done) {
+              throw "extraneous data";
+            }
```

Comment:
The logic of `receivedTrailers` doesn't seem solid to me. Trailers are crucial in gRPC because they contain the status, including the successful `grpc-status: 0`. It is a protocol error if we don't receive a grpc-status, and it appears that we're just ignoring this case.

The documentation states that trailers are "populated at the 'end' event". We can make use of that, a create a promise for the trailer early on, same as we do in the http2 gRPC transport. This allows us to bring the two implementations closer together:

```suggestion
            const responseHeaders = nodeHeaderToWebHeader(response.headers);
            const trailerPromise = readHttp1ResponseTrailer(response);
            assert(
              typeof response.statusCode == "number",
              "http1 client response is missing status code"
            );
            validateResponse(
              useBinaryFormat,
              response.statusCode,
              responseHeaders
            );
            const messageResult = await readEnvelope(response);
            const eofResult = await readEnvelope(response);
            if (!eofResult.done) {
              throw "extraneous data";
            }
            const trailer = await trailerPromise;
            validateGrpcStatus(trailer);
            if (messageResult.done) {
              throw "premature eof";
            }
```

The function readHttp1ResponseTrailer (can put this next to readResponseTrailer in io.ts):

```ts
/**
 * Returns a promise for HTTP/1 response trailers.
 */
export function readHttp1ResponseTrailer(
  response: http.IncomingMessage
): Promise<Headers> {
  return new Promise<Headers>((resolve, reject) => {
    if (response.errored) {
      return reject(response.errored);
    }
    if (response.readableEnded) {
      return resolve(nodeHeaderToWebHeader(response.trailers));
    }
    response.once("error", error);
    response.once("end", end);

    function end() {
      response.off("error", error);
      response.off("end", end);
      resolve(nodeHeaderToWebHeader(response.trailers));
    }

    function error(err: Error) {
      response.off("error", error);
      response.off("end", end);
      reject(err);
    }
  });
}
```

Reply:
See https://github.com/bufbuild/connect-web/pull/368 for a change to the http2 transport that takes the same code path as suggested above.